controllers/runner_controller.go

```diff
@@ -927,3 +944,82 @@ func workVolumeMountPresent(items []corev1.VolumeMount) (bool, int) {
 	}
 	return false, 0
 }
+
+func applyWorkVolumeClaimTemplate(runner *v1alpha1.Runner) error {
+	isPresent, _ := workVolumeMountPresent(runner.Spec.VolumeMounts)
+	if isPresent {
+		return errors.New("volumeMount with name \"work\" should not exist if workVolumeClaimTemplate is specified")
+	}
+
+	isPresent, _ = workVolumePresent(runner.Spec.Volumes)
+	if isPresent {
+		return errors.New("volume with name \"work\" should not exist if workVolumeClaimTemplate is specified")
+	}
+
+	runner.Spec.VolumeMounts = append(runner.Spec.VolumeMounts, corev1.VolumeMount{
+		MountPath: "/runner/_work",
+		Name:      "work",
+	})
+
+	runner.Spec.Volumes = append(runner.Spec.Volumes, corev1.Volume{
+		Name: "work",
+		VolumeSource: corev1.VolumeSource{
+			Ephemeral: &corev1.EphemeralVolumeSource{
+				VolumeClaimTemplate: &corev1.PersistentVolumeClaimTemplate{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: runner.ObjectMeta.Name + "-work",
+					},
+					Spec: corev1.PersistentVolumeClaimSpec{
+						AccessModes:      runner.Spec.WorkVolumeClaimTemplate.AccessModes,
+						StorageClassName: &runner.Spec.WorkVolumeClaimTemplate.StorageClassName,
+						Resources:        runner.Spec.WorkVolumeClaimTemplate.Resources,
+					},
+				},
+			},
+		},
+	})
+	return nil
+}
+
+// isRequireSameNode specifies for the runner in kubernetes mode wether it should
+// schedule jobs to the same node where the runner is
+func isRequireSameNode(pod *corev1.Pod) (bool, error) {
+	isPresent, index := workVolumePresent(pod.Spec.Volumes)
+	if !isPresent {
+		return true, nil
+	}
+
+	if pod.Spec.Volumes[index].Ephemeral == nil || pod.Spec.Volumes[index].Ephemeral.VolumeClaimTemplate == nil {
+		// TODO(nikola-jokic): how to handle this case?
+		return true, nil
+	}
+
+	for i := range pod.Spec.Volumes[index].Ephemeral.VolumeClaimTemplate.Spec.AccessModes {
+		switch pod.Spec.Volumes[index].Ephemeral.VolumeClaimTemplate.Spec.AccessModes[i] {
+		case corev1.ReadWriteOnce:
+			return true, nil
+		case corev1.ReadWriteMany:
+		default:
+			// TODO (nikola-jokic): Should we throw error or delete it from access modes?
+			return true, errors.New("read only mode cannot be specified for actions-runner-controller to work properly")
+		}
+	}
+	return false, nil
+}
+
+func appendRunnerVolumeMountEnvs(pod *corev1.Pod) error {
+	setRunnerEnv(pod, "ACTIONS_RUNNER_CLAIM_NAME", pod.ObjectMeta.Name+"-work")
```

Comment:
What name should I set if some user decides to create a volume by hand?
Should we not allow work volumes in `containerMode: kubernetes` ?

Reply:
I think we should just fail if volumes or volume mounts from the user interact with the ones we are trying to set.